### PR TITLE
Fix Stack overflow in sql planning in debug builds

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?
Closes https://github.com/apache/arrow-datafusion/issues/4065 from @andygrove 
Closes https://github.com/apache/arrow-datafusion/issues/4728 from @Jefffrey 

Also reported by @maxburke  in slack

# Rationale for this change
Debug builds are overflowing the stack causing pain and suffering for users. 

# What changes are included in this PR?

1. Extract several parts of the body of `SqlToRel::sql_expr_to_logical_expr` into separate functions

The reason this reduces stack size in debug builds is explained in the "Technical Backstory" heading of https://github.com/apache/arrow-datafusion/pull/1047

# Are these changes tested?
Covered by existing coverage

# Are there any user-facing changes?
None intended (other than avoiding stack overflows)

### CODE REVIEW

File: datafusion/sql/src/planner.rs [hunk 1]
```diff
@@ -1866,31 +1866,7 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
-                    // compound identifiers, but this is not a compound
-                    // identifier. (e.g. it is "foo.bar" not foo.bar)
-
-                    Ok(Expr::Column(Column {
-                        relation: None,
-                        name: normalize_ident(id),
-                    }))
-                }
-            }
+            SQLExpr::Identifier(id) => self.sql_identifier_to_expr(id),
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:1869]: I just applied this pattern several times: extract code into a separate function

File: datafusion/sql/src/planner.rs [hunk 2]
```diff
@@ -2193,159 +2038,11 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
-                    Expr::AggregateFunction(expr::AggregateFunction {
-                        fun, args, distinct, ..
-                    }) => Ok(Expr::AggregateFunction(expr::AggregateFunction::new( fun, args, distinct, Some(Box::new(self.sql_expr_to_logical_expr(*filter, schema, planner_context)?)) ))),
-                    _ => Err(DataFusionError::Internal("AggregateExpressionWithFilter expression was not an AggregateFunction".to_string()))
-                }
-            }
+            SQLExpr::AggregateExpressionWithFilter { expr, filter } => self.sql_agg_with_filter_to_expr(expr, filter, schema, planner_context),
 
-            SQLExpr::Function(mut function) => {
-                let name = if function.name.0.len() > 1 {
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:2230]: I think the expanded handling of `SQLExpr::Function` is likely the thing that resulted in significantly larger stacks and thus overflows

File: datafusion/sql/src/planner.rs [hunk 3]
```diff
@@ -2375,6 +2072,395 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         }
     }
 
+    fn sql_function_to_expr(
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:2075]: All this code was simply moved from above

**Review by andygrove — APPROVED** [review]
TIL... thanks @alamb!

**Comment by alamb** [comment]
https://github.com/apache/arrow-datafusion/actions/runs/3812581550/jobs/6485893384

😢  still overflows on windows

```
test tpcds_physical_q70 ... ok
test tpcds_physical_q71 ... ok
test tpcds_physical_q72 ... ok
test tpcds_physical_q73 ... ok

thread 'tpcds_physical_q64' has overflowed its stack
error: test failed, to rerun pass `-p datafusion --test tpcds_planning`
```

File: datafusion/core/tests/tpcds_planning.rs [hunk 4]
```diff
@@ -855,7 +850,6 @@ async fn tpcds_physical_q63() -> Result<()> {
     create_physical_plan(63).await
 }
 
-#[ignore] // thread 'q64' has overflowed its stack
 #[tokio::test]
```
Comment (alamb) [inline datafusion/core/tests/tpcds_planning.rs:853]:
```suggestion
#[ignore] // thread 'q64' has overflowed its stack
#[tokio::test]
```

File: datafusion/core/tests/tpcds_planning.rs [hunk 5]
```diff
@@ -345,7 +343,6 @@ async fn tpcds_logical_q63() -> Result<()> {
     create_logical_plan(63).await
 }
 
-#[ignore] // thread 'q64' has overflowed its stack
 #[tokio::test]
```
Comment (alamb) [inline datafusion/core/tests/tpcds_planning.rs:346]:
```suggestion
#[ignore] // thread 'q64' has overflowed its stack
```

File: datafusion/core/tests/tpcds_planning.rs [hunk 6]
```diff
@@ -346,7 +344,6 @@ async fn tpcds_logical_q63() -> Result<()> {
 }
 
 #[ignore] // thread 'q64' has overflowed its stack
```
Comment (alamb) [inline datafusion/core/tests/tpcds_planning.rs:346]:
```suggestion
#[ignore] // thread 'q64' has overflowed its stack]
#[tokio::test]
```

**Comment by alamb** [comment]
Looks like q64 needs some more work -- will do so

**Comment by alamb** [comment]
I have one more PR that should avoid stack overflows while planning (SQL --> LogicalPlan) but then planning overflows at a later step -- see https://github.com/apache/arrow-datafusion/issues/4786

**Pull request merged**